resources/test/nested-testharness.js

```diff
@@ -0,0 +1,78 @@
+'use strict';
+
+/**
+ * Execute testharness.js and one or more scripts in an iframe. Report the
+ * results of the execution.
+ *
+ * @param {...function|string} bodies - a function body. If specified as a
+ *                                      function object, it will be converted
+ *                                      to a string using the built-in
+ *                                      `Function.prototype.toString` prior to
+ *                                      inclusion in the generated iframe.
+ *
+ * @returns {Promise} eventual value describing the result of the test
+ *                    execution; the summary object has two properties:
+ *                    `harness` (a string describing the harness status) and
+ *                    `tests` (an object whose "own" property names are the
+ *                    titles of the defined sub-tests and whose associated
+ *                    values are the subtest statuses).
+ */
+function makeTest(...bodies) {
+  const closeScript = '<' + '/script>';
+  let src = `
+<!DOCTYPE HTML>
+<html>
+<head>
+<title>Document title</title>
+<script src="/resources/testharness.js?${Math.random()}">${closeScript}
+</head>
+
+<body>
+<div id="log"></div>`;
+  bodies.forEach((body) => {
+    src += '<script>(' + body + ')();' + closeScript;
+  });
+
+  const iframe = document.createElement('iframe');
+
+  document.body.appendChild(iframe);
+  iframe.contentDocument.write(src);
+
+  return new Promise((resolve) => {
+    window.addEventListener('message', function onMessage(e) {
+      if (e.source !== iframe.contentWindow) {
+        return;
+      }
+      if (!e.data || e.data.type !=='complete') {
+        return;
+      }
+      window.removeEventListener('message', onMessage);
```

Comment:
Since the event listener will at most invoke `resolve` again which won't do anything, you could simplify this by just using `{once: true}` as the 3rd argument to `addEventListener`. Even interpreting the options object as a true `capture` argument in an older browser wouldn't matter because the event target for this even is window.

Reply:
Works for me

Reply:
It turns out this doesn't work for me :P Since that's testharness.js running in there, we receive a bunch of messages before the one we're expecting. If we use the DOM API to unbind after the first, this handler won't be executed when the expected event is emitted.

Reply:
Aha, makes sense!